Initial implementation for `DssLitePsm` (#1)
makerdao/dss-lite-psm: README.md

- Mainnet: \[TBD\]
- Goerli: \[TBD\]

## Overview

A Peg Stability Module (PSM) is a facility through which users can freely swap Dai for stablecoins with no slippage.
MakerDAO Governance can enable swap fees, though, which are computed as revenue for the protocol.

This module is heavily inspired by the [current PSM][psm], [PSM v2][psm-v2] and some other PSM prototypes within
MakerDAO repositories.

The issue with those implementations is that swapping through them can be quite gas intensive, because they manipulate
the `Vat` (MakerDAO's main accounting module) directly on every swap.

To help alleviate this problem, `DssLitePsm` aims to be more gas efficient. The strategy is to allow users to swap in a

> **pool** of pre-minted Dai and stablecoins, reducing the base case of a swap to 2 ERC-20 token transfers with little

Author: Done

> **pool** of pre-minted Dai and stablecoins, reducing the base case of a swap to 2 ERC-20 token transfers with little

Reviewer: LGTM

> **pool** of pre-minted Dai and stablecoins, reducing the base case of a swap to 2 ERC-20 token transfers with little

Reviewer 2: LGTM, resolving

functions to front-run them, causing the swap to fail, as the Dai liquidity would be lower than the required amount.

The scenario A above is not possible with the current PSM implementation because each swap is "self-balancing", so no
off-band bookkeeping is required.

**Scenario B:** a large swap might front-run another one, even if unintentionally. Imagine there is `10M` Dai
outstanding in `DssLitePsm`. If Alice &ndash; who wants to swap `8M` &ndash; and Bob &ndash; who wants to swap `3M`
&ndash; submit their transactions at the same time, only the first one will be executed.

The scenario B above is not possible with the current PSM implementation because `sellGem` is able to mint Dai
on-the-fly to fulfill the swap, given that there is enough room in the debt ceiling.

Notice how the same issue happens in `buyGem`, however the amount of `gem` deposited into `DssLitePsm` is only bounded
by the debt ceiling, while the amount of `Dai` will tend to gravitate towards `buf`.

The consequence is that anyone willing to call `sellGem` with a value larger than `buf` should take care of potential
front-running transactions by bundling it with an optional liquidity increase (`fill`).

> 

Reviewer: LGTM excepting a minor modification in a line.

> 

Reviewer: LGTM

> 

Reviewer: Modification looks good as well.

> 

Reviewer 2: LGTM, resolving

> 

Reviewer 2: LGTM, resolving

#### 2. Emergency Shutdown

> `DssLitePsm` completely disregards Emergency Shutdown. This is intentional, as it complicates a lot the design.

Author: Done

> `DssLitePsm` completely disregards Emergency Shutdown. This is intentional, as it complicates a lot the design.

Reviewer: LGTM

> `DssLitePsm` completely disregards Emergency Shutdown. This is intentional, as it complicates a lot the design.

Reviewer 2: LGTM, resolving

## Contributing

> TODO.

Author: Added the relevant section.

> TODO.

Reviewer: LGTM

only constraint is that `DssLitePsm` **should** be able to freely move any amount of `gem` on behalf of such address.

## Architecture

A simplified diagram of the `DssLitePsm` architecture:

```
                                                buyGemNoFee /
                                           ╭────────────────────────  🤴
 ╭─────────╮                               │    sellGemNoFee      Whitelisted
 │         │       transferFrom            │                         User
 │   Gem   ◄────────────────────────────╮  │
 │         │                            │  │
 ╰────▲────╯                            │  │
      │                                 │  │
      │                                 │  │
      │ approve ·····╮                  │  │
      │              ╎                  │  │
╭─────┴────╮         ╎          ╭───────┴──▼───╮
│          │         ╎          │              │     buyGem /
│  Pocket  │         ╰·········>│  DssLitePsm  ◄────────────────────  🧑
│          │                    │              │     sellGem         User
╰──────────╯                    ╰─┬──┬──┬──▲───╯

>                                   │  │  │  │
>                                   │  │  │  │
>               slip / frob         │  │  │  │
>      ╭────────────────────────────╯  │  │  │
>      │                               │  │  │

Author: `slip` is no longer called after the latest refactor.
```suggestion
                                  │  │  │  │
                                  │  │  │  │
                  frob            │  │  │  │
     ╭────────────────────────────╯  │  │  │
     │                               │  │  │
```

Reviewer: Yeah looks good, let's leave this one as well for the round after audit feedback.

Author: Yeah, I'm just adding it here so I don't forget.

The consequence is that anyone willing to call `sellGem` with a value larger than `buf` should take care of potential
front-running transactions by bundling it with an optional liquidity increase (`fill`).

#### 2. No Slippage Protection

Swaps in `DssLitePsm` are generally not subject to slippage. The only exception is when there is a MakerDAO Governance

> proposal do increase the swapping fees `tin` and `tout`. That is done through an Executive Spell, which is an on-chain

Reviewer: ```suggestion
proposal to increase the swapping fees `tin` and/or `tout`. That is done through an Executive Spell, which is an on-chain
```

smart contract that can be permissionlessly _cast_ (executed) after following the Governance process.

If Alice sends a swap transaction and a spell increasing the fees is cast before her transaction, she will either pay
more Dai when buying gems or receive less Dai when selling gems than the originally expected.

This is a highly unlikely scenario, but users or aggregators are able to handle this issue through a wrapper contract.

#### 3. No Support for Upgradeable Gems

We no longer have a dedicated `GemJoin` contract to normalize different token implementations. For instance, we lost the
capacity to identify upgrades in upgradeable tokens when compared to the previous iteration of the [PSM][gem-join-8].

On the other hand, non-upgradeable gems that [do not return `true` on `transfer`/`transferFrom`][weird-erc20] were not
previously supported, but we removed such restriction in this iteration.

> 

Reviewer: ```suggestion
```
